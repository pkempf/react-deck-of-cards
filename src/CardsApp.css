.CardsApp {
  margin-top: 20px;
}

.CardsApp-pile {
  margin: 10px auto;
  padding: 80px;
  display: fixed;
}

.CardsApp-draw-btn {
  padding: 10px;
  border: 3px solid black;
  border-radius: 8px;
  background: linear-gradient(to bottom, #333333, #000000);
  color: white;
  cursor: pointer;
}

.CardsApp-draw-btn:hover {
  background: linear-gradient(to bottom, #555555, #333333);
}
