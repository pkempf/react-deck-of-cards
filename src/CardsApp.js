import React, { useState, useEffect, useRef } from "react";
import axios from "axios";
import Card from "./Card";
import "./CardsApp.css";

const BASE_URL = "http://deckofcardsapi.com/api/deck";

const CardsApp = () => {
  const [deck, setDeck] = useState(null);
  const [drawnCards, setDrawnCards] = useState([]);
  const [auto, setAuto] = useState(false);
  const [buttonMsg, setButtonMsg] = useState("Draw some cards!");
  const timerRef = useRef(null);

  // on load, get the deck from the api
  useEffect(() => {
    async function getData() {
      let res = await axios.get(`${BASE_URL}/new/shuffle`);
      setDeck(res.data);
    }
    getData();
  }, [setDeck]);

  // drawing cards from the api, if enamled
  useEffect(() => {
    // get a card from the api and add it to state drawnCards
    async function getCard() {
      let { deck_id } = deck;

      try {
        let res = await axios.get(`${BASE_URL}/${deck_id}/draw/`);

        // if the api call tells us no cards remaining, we're done
        if (res.data.remaining === 0) {
          setAuto(false);
          setButtonMsg("No cards left!");
          return;
        }

        // get the actual card out of res
        const drawnCard = res.data.cards[0];

        // put it in state
        setDrawnCards((drawnCards) => [
          ...drawnCards,
          {
            id: drawnCard.code,
            name: drawnCard.value + " of " + drawnCard.suit,
            imgSrc: drawnCard.image,
          },
        ]);
      } catch (err) {
        alert(err);
      }
    }

    // if we're autodrawing but we don't have a timer, set one
    if (auto && !timerRef.current) {
      timerRef.current = setInterval(async () => {
        await getCard();
      }, 1000);
    }

    // cleanup - stop the timer and remove it from ref
    return () => {
      clearInterval(timerRef.current);
      timerRef.current = null;
    };
  }, [auto, setAuto, deck]);

  const toggleAuto = () => {
    setAuto((auto) => !auto);
    if (!auto) setButtonMsg("Stop drawing cards!");
    else setButtonMsg("Draw some cards!");
  };

  const cardsToRender = drawnCards.map((card) => (
    <Card id={card.id} name={card.name} imgSrc={card.imgSrc} />
  ));

  return (
    <div className="CardsApp">
      {deck ? (
        <button className="CardsApp-draw-btn" onClick={toggleAuto}>
          {buttonMsg}
        </button>
      ) : null}
      <div className="CardsApp-pile">{cardsToRender}</div>
    </div>
  );
};

export default CardsApp;
