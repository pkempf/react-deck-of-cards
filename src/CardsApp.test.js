import { render } from "@testing-library/react";
import CardsApp from "./CardsApp";

it("renders without crashing", () => {
  render(<CardsApp />);
});

it("matches the snapshot", () => {
  const { asFragment } = render(<CardsApp />);
  expect(asFragment()).toMatchSnapshot();
});
