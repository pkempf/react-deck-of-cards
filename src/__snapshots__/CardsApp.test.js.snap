// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="CardsApp"
  >
    <div
      class="CardsApp-pile"
    />
  </div>
</DocumentFragment>
`;
